vpc/service/assign_addresses_v3.go

```diff
@@ -98,27 +102,56 @@ type branchENI struct {
 	idx           int
 }
 
+func subnetCacheKey(az, accountID string, subnetIDs []string) string {
+	sort.Strings(subnetIDs)
+	return fmt.Sprintf("az:%s accountID:%s subnetIDs:%s", az, accountID, strings.Join(subnetIDs, ","))
+}
+
 func (vpcService *vpcService) getSubnet(ctx context.Context, az, accountID string, subnetIDs []string) (*subnet, error) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	ctx, span := trace.StartSpan(ctx, "getSubnet")
-	defer span.End()
+	// NOTE[jigish] caching using the full set of subnetIDs will cause the same subnet to be returned for all hits
+	cacheKey := subnetCacheKey(az, accountID, subnetIDs)
+	r, err, _ := vpcService.getSubnetLock.Do(cacheKey, func() (interface{}, error) {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		ctx, span := trace.StartSpan(ctx, "getSubnet")
+		defer span.End()
+
+		item := vpcService.getSubnetCache.Get(cacheKey)
+		var staleSubnet *subnet
+		if item != nil {
+			span.AddAttributes(trace.BoolAttribute("cached", true))
+			span.AddAttributes(trace.BoolAttribute("expired", item.Expired()))
+			span.AddAttributes(trace.StringAttribute("expires", item.Expires().String()))
+
+			val := item.Value().(*subnet)
+			if !item.Expired() {
+				return val, nil
+			} else if time.Now().Before(item.Expires().Add(subnetStaleValidTime)) {
+				staleSubnet = val
```

Comment:
Thinking back to it now, I do not see any reason not to always serve stale subnets.

Reply:
deleting subnets?

Reply:
If we're unable to query the database, a subnet has been deleted, and we cannot restart the service, it's a bad day and we probably have other problems.

Reply:
ok removed the valid time check